cudax/include/cuda/experimental/buffer

```diff
@@ -0,0 +1,204 @@
+//!    resource exceeds the lifetime of the buffer.
+//!
+//! @endrst
+template <class _Tp, class... _Properties>
+class uninitialized_async_buffer
+{
+private:
+  _CUDA_VMR::async_resource_ref<_Properties...> __mr_;
+  ::cuda::stream_ref __stream_ = {};
+  size_t __count_              = 0;
+  void* __buf_                 = nullptr;
+
+  //! @brief Determines the allocation size given the alignment and size of @p _Tp
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY static constexpr size_t
+  __get_allocation_size(const size_t __count) noexcept
+  {
+    constexpr size_t __alignment = alignof(_Tp);
+    return (__count * sizeof(_Tp) + (__alignment - 1)) & ~(__alignment - 1);
+  }
+
+  //! @brief Determines the properly aligned start of the buffer given the alignment and size of @p _Tp
+  _CCCL_NODISCARD _LIBCUDACXX_INLINE_VISIBILITY constexpr _Tp* __get_data() const noexcept
+  {
+    constexpr size_t __alignment = alignof(_Tp);
+    size_t __space               = __get_allocation_size(__count_);
+    void* __ptr                  = __buf_;
+    return reinterpret_cast<_Tp*>(_CUDA_VSTD::align(__alignment, __count_ * sizeof(_Tp), __ptr, __space));
+  }
+
+public:
+  //! @brief Constructs a ``cuda::uninitialized_async_buffer`` from a memory resource and a number of elements
+  //! @param __mr The memory resource to allocate the buffer with.
+  //! @param __count The desired size of the buffer.
+  uninitialized_async_buffer(
+    _CUDA_VMR::async_resource_ref<_Properties...> __mr, const ::cuda::stream_ref __stream, const size_t __count)
+      : __mr_(__mr)
+      , __stream_(__stream)
+      , __count_(__count)
+      , __buf_(__mr_.allocate_async(__get_allocation_size(__count_), __stream_))
+  {}
+
+  uninitialized_async_buffer(const uninitialized_async_buffer&)            = delete;
+  uninitialized_async_buffer& operator=(const uninitialized_async_buffer&) = delete;
+  uninitialized_async_buffer(uninitialized_async_buffer&& __other) noexcept
+      : __mr_(__other.__mr_)
+      , __stream_(__other.__stream_)
+      , __count_(__other.__count_)
+      , __buf_(__other.__buf_)
+  {
+    __other.__stream_ = {};
+    __other.__count_  = 0;
+    __other.__buf_    = nullptr;
+  }
+  uninitialized_async_buffer& operator=(uninitialized_async_buffer&& __other) noexcept
+  {
+    if (__buf_)
+    {
+      __stream_.wait();
+      __mr_.deallocate_async(__buf_, __get_allocation_size(__count_), __stream_);
+    }
+    __mr_             = __other.__mr_;
+    __stream_         = __other.__stream_;
+    __count_          = __other.__count_;
+    __buf_            = __other.__buf_;
+    __other.__stream_ = {};
+    __other.__count_  = 0;
+    __other.__buf_    = nullptr;
+    return *this;
+  }
+
+  //! @brief Destroys a ``cuda::uninitialized_async_buffer`` deallocating the buffer
+  ~uninitialized_async_buffer()
+  {
+    if (__buf_)
+    {
+      __stream_.wait();
```

Comment:
@jrhemstad : that is too expensive we should not do that

Reply:
Agree. Caller is responsible for synchronizing or keeping the buffer alive as necessary.